### PR TITLE
[VKCI-90] common code for Kubernetes CRUD operation

### PR DESCRIPTION
**Pull request by ymo24**
* common code for Kubernetes CRUD operation

Signed-off-by: ymo24 <ymo@vmware.com>

- - -
This change is [Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)

### CODE REVIEW

**Review by lzichong — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)* status: 0 of 2 files reviewed, 8 unresolved discussions (waiting on @arunmk, @rliang88, and @ymo24)

___
*[`pkg/Testingsdk/k8sclient.go` line 21 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo-flxEd-Ku0hG1d7x:-NIo-flxEd-Ku0hG1d7y:b-ij5mrb) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L21)):*
> ```Go
> )
> 
> var (
> ```

nit: const



_Code quote:_
```Go
var
```
___
*[`pkg/Testingsdk/k8sclient.go` line 24 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo1DGy1wmIE-6giZj8:-NIo1DGy1wmIE-6giZj9:bqo1t4f) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L24)):*
> ```Go
> 	ResourceExisted   = errors.New("[REX] resource is already existed")
> 	ResourceNotFound  = errors.New("[RNF] resource is not found")
> 	ResourceNameNulll = errors.New("[RNN] resource name is null")
> ```

typo: ResourceNameNull



_Code quote:_
```Go
ResourceNameNulll
```
___
*[`pkg/Testingsdk/k8sclient.go` line 34 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo-V2g1ID3K0cTaYkM:-NIo-V2g1ID3K0cTaYkN:b63md1f) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L34)):*
> ```Go
> 		pvc, err := getPVC(ctx, k8sClient, nameSpace, pvcName)
> 		if err != nil {
> 			if apierrs.IsInternalError(err) || apierrs.IsTimeout(err) || apierrs.IsServerTimeout(err) ||
> ```

Could we move all usage of this check into a function called `IsRetryableError(err) bool`?



_Code quote:_
```Go
			if apierrs.IsInternalError(err) || apierrs.IsTimeout(err) || apierrs.IsServerTimeout(err) ||
				apierrs.IsTooManyRequests(err) || utilnet.IsProbableEOF(err) || utilnet.IsConnectionReset(err) {
				return false, nil
			}
```
___
*[`pkg/Testingsdk/k8sclient.go` line 43 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo07Ig6CA0-Tx4U21z:-NIo07Ig6CA0-Tx4U22-:bwkepbc) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L43)):*
> ```Go
> 			return false, nil
> 		}
> 		if pvc != nil && pvc.Status.Phase == "Bound" {
> ```

Is there a apiv1 constant for this similar to `apiv1.PodRunning`?



_Code quote:_
```Go
"Bound"
```
___
*[`pkg/Testingsdk/k8sclient.go` line 60 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo-y6-9RIFm52hFkk8:-NIo-y6-9RIFm52hFkk9:b-othm7e) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L60)):*
> ```Go
> func isDeploymentReady(ctx context.Context, k8sClient *kubernetes.Clientset, nameSpace string, deployName string) (bool, error) {
> 	retryInterval := 20 * time.Second
> 	retryTimeout := 140 * time.Second
> ```

Can we move these to as constants? `defaultRetryInterval` and `defaultRetryTimeout`?



_Code quote:_
```Go
	retryInterval := 20 * time.Second
	retryTimeout := 140 * time.Second
```
___
*[`pkg/Testingsdk/k8sclient.go` line 189 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo0j8bDhig1hm3tYZ0:-NIo0j8bDhig1hm3tYZ1:bfn72k8) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L189)):*
> ```Go
> 		_, err := k8sClient.CoreV1().Namespaces().Get(ctx, nameSpace, metav1.GetOptions{})
> 		if err != nil {
> 			if err.(*apierrs.StatusError).ErrStatus.Reason == "NotFound" {
> ```

Can we use use an apierrs here, with `apierrs.IsNotFound(err)` instead of checking by "NotFound"? Similar to all the other areas where we are using this check



_Code quote:_
```Go
			if err.(*apierrs.StatusError).ErrStatus.Reason == "NotFound" {
				return true, nil
			}
```
___
*[`pkg/Testingsdk/k8sclient.go` line 243 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo12EC77DJqlcMOMIF:-NIo12EC77DJqlcMOMIG:b48k95r) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L243)):*
> ```Go
> }
> 
> func getPodByPrefix(ctx context.Context, k8sClient *kubernetes.Clientset, nameSpace string, podNamePrefix string) ([]*apiv1.Pod, error) {
> ```

nit: getPodsByPrefix as we are returning a list of pods



_Code quote:_
```Go
getPodByPrefix
```
___
*[`pkg/Testingsdk/client.go` line 1 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo-9Oq45bLSfJLaGAc:-NIo-9Oq45bLSfJLaGAd:bknpmsn) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/client.go#L1)):*
> ```Go
> package Testingsdk
> ```

Can we replace this to `testingsdk` with lowercase to keep same style as the others?

**Review by ymo24 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @arunmk, @lzichong, and @rliang88)

___
*[`pkg/Testingsdk/client.go` line 1 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo-9Oq45bLSfJLaGAc:-NIs_bxl0JdLhKpHmMqv:b-896fix) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/client.go#L1)):*
<details><summary><i>Previously, lzichong wrote…</i></summary><blockquote>

Can we replace this to `testingsdk` with lowercase to keep same style as the others?
</blockquote></details>

Done.
___
*[`pkg/Testingsdk/k8sclient.go` line 34 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo-V2g1ID3K0cTaYkM:-NIs_ja40THVSTqYYtr2:b-896fix) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L34)):*
<details><summary><i>Previously, lzichong wrote…</i></summary><blockquote>

Could we move all usage of this check into a function called `IsRetryableError(err) bool`?
</blockquote></details>

Done.



_Code quote:_
```Go
apierrs.IsTooManyRequests(err) || utilnet.IsProbableEOF(err)
```
___
*[`pkg/Testingsdk/k8sclient.go` line 43 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo07Ig6CA0-Tx4U21z:-NIs_fIWCgjRuP48HNg2:b-896fix) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L43)):*
<details><summary><i>Previously, lzichong wrote…</i></summary><blockquote>

Is there a apiv1 constant for this similar to `apiv1.PodRunning`?
</blockquote></details>

Done.
___
*[`pkg/Testingsdk/k8sclient.go` line 189 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo0j8bDhig1hm3tYZ0:-NIs_YKv4sXkSqVfM_US:brxg4wi) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L189)):*
> "NotFound"
Already updated, thanks Zichong

**Review by ymo24 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @arunmk, @lzichong, and @rliang88)

___
*[`pkg/Testingsdk/k8sclient.go` line 21 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo-flxEd-Ku0hG1d7x:-NIsgmZLA3n-4axBrYF7:b7qk8hq) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L21)):*
<details><summary><i>Previously, lzichong wrote…</i></summary><blockquote>

nit: const
</blockquote></details>

We can keep error as var. Thanks Zichong
___
*[`pkg/Testingsdk/k8sclient.go` line 60 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIo-y6-9RIFm52hFkk8:-NIsgrKRB1n-ovoeRS2Q:b-896fix) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L60)):*
<details><summary><i>Previously, lzichong wrote…</i></summary><blockquote>

Can we move these to as constants? `defaultRetryInterval` and `defaultRetryTimeout`?
</blockquote></details>

Done.

**Review by lzichong — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)* status: 1 of 2 files reviewed, 6 unresolved discussions (waiting on @arunmk and @rliang88)

**Review by lzichong — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)* status: 1 of 3 files reviewed, 8 unresolved discussions (waiting on @arunmk, @rliang88, and @ymo24)

___
*[`pkg/testingsdk/client.go` line 276 at r4](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NJ7cZCCE_eLRTnsKqx2:-NJ7cZCCE_eLRTnsKqx3:b-epnp6x) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/cb9c8c110e6067cc333a34cf0570e39b662b437a/pkg/testingsdk/client.go#L276)):*
> ```Go
> }
> 
> func (tc *TestClient) IsWorkerNodeNotReady(ctx context.Context, workerNode *apiv1.Node) error {
> ```

Thanks for implementing this! Could you also add a comment why we can't use negate result from `IsWorkerNodeReady()`?



_Code quote:_
```Go
func (tc *TestClient) IsWorkerNodeNotReady(ctx context.Context, workerNode *apiv1.Node) error {
```

**Review by rliang88 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)* status: 1 of 3 files reviewed, 9 unresolved discussions (waiting on @arunmk, @lzichong, and @ymo24)

___
*[`pkg/Testingsdk/k8sclient.go` line 24 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NJ7cuBQAgsIgJmorq8U:-NJ7cuBQAgsIgJmorq8V:b-4154i2) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L24)):*
> ```Go
> 	ResourceExisted   = errors.New("[REX] resource is already existed")
> 	ResourceNotFound  = errors.New("[RNF] resource is not found")
> 	ResourceNameNulll = errors.New("[RNN] resource name is null")
> ```

3 l's in this variable name, is this intentional?

**Review by ymo24 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)* status: 1 of 3 files reviewed, 9 unresolved discussions (waiting on @arunmk, @lzichong, and @rliang88)

___
*[`pkg/Testingsdk/k8sclient.go` line 24 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NJ7cuBQAgsIgJmorq8U:-NJ7dY96707mJ6_5-YLc:bhof8fn) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L24)):*
<details><summary><i>Previously, rliang88 (Richard Liang) wrote…</i></summary><blockquote>

3 l's in this variable name, is this intentional?
</blockquote></details>

Are you looking at the latest version?

**Review by rliang88 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)* status: 1 of 3 files reviewed, 8 unresolved discussions (waiting on @arunmk, @lzichong, and @ymo24)

___
*[`pkg/Testingsdk/k8sclient.go` line 24 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NJ7cuBQAgsIgJmorq8U:-NJ7eJ-u75viTxtN-epD:bemrzxu) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/8243531b99fdf5c4c89a676a419340e956d83518/pkg/Testingsdk/k8sclient.go#L24)):*
<details><summary><i>Previously, ymo24 wrote…</i></summary><blockquote>

Are you looking at the latest version?
</blockquote></details>

oh good, it changed

**Review by rliang88 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-:-NJ7el5pEqB3H-_uXlNp:bnfp4nl)* status: 1 of 3 files reviewed, 8 unresolved discussions (waiting on @arunmk, @lzichong, and @ymo24)

**Review by lzichong — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r2, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-:-NJBl_lo-HBKmh-HIkFd:bnfp4nl)* status: 2 of 3 files reviewed, 8 unresolved discussions (waiting on @arunmk and @ymo24)

File: pkg/testingsdk/client.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("error creating Storage Class [%s] for cluster [%s]: [%v]", scName, tc.ClusterId, err)
+	}
+	return sc, nil
+}
+
+func (tc *TestClient) CreatePV(ctx context.Context, persistentVolumeName string, storageClass string, storageProfile string, storageSize string, reclaimPolicy apiv1.PersistentVolumeReclaimPolicy) (*apiv1.PersistentVolume, error) {
+	pv, err := createPV(ctx, tc.Cs.(*kubernetes.Clientset), persistentVolumeName, storageClass, storageProfile, storageSize, reclaimPolicy)
+	if err != nil {
+		return nil, fmt.Errorf("error creating Persistent Volume [%s] for cluster [%s]: [%v]", persistentVolumeName, tc.ClusterId, err)
```
Comment (arunmk) [inline pkg/testingsdk/client.go:68]: clusterName would be more helpful in these logs while debugging

File: pkg/testingsdk/client.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+func (tc *TestClient) IsDeploymentReady(ctx context.Context, nameSpace string, deployName string) error {
+	err := isDeploymentReady(ctx, tc.Cs.(*kubernetes.Clientset), nameSpace, deployName)
+	if err != nil {
+		return fmt.Errorf("error querying Deployment [%s] status for cluster [%s]: [%v]", deployName, tc.ClusterId, err)
+	}
+	return nil
+}
+
+func (tc *TestClient) IsWorkerNodeReady(ctx context.Context, workerNode *apiv1.Node) error {
+	err := wait.PollImmediate(defaultNodeInterval, defaultNodeReadyTimeout, func() (bool, error) {
```
Comment (arunmk) [inline pkg/testingsdk/client.go:229]: nice!

File: pkg/testingsdk/k8sclient.go [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+	defaultRetryInterval       = 10 * time.Second
+	defaultRetryTimeout        = 160 * time.Second
+	defaultLongRetryInterval   = 20 * time.Second
+	defaultLongRetryTimeout    = 300 * time.Second
+	defaultNodeInterval        = 2 * time.Second
+	defaultNodeReadyTimeout    = 20 * time.Minute
+	defaultNodeNotReadyTimeout = 8 * time.Minute
+)
+
+func isPvcReady(ctx context.Context, k8sClient *kubernetes.Clientset, nameSpace string, pvcName string) error {
```
Comment (arunmk) [inline pkg/testingsdk/k8sclient.go:36]: this is actually a `waitForPvcReady`, correct?

File: pkg/testingsdk/k8sclient.go [hunk 4]
```diff
@@ -0,0 +1,652 @@
...
+		if !ready {
+			fmt.Printf("pvc %s is not bound\n", pvc.Name)
+			return false, nil
+		}
+		return true, nil
+	})
+	return err
+}
+
+func isDeploymentReady(ctx context.Context, k8sClient *kubernetes.Clientset, nameSpace string, deployName string) error {
```
Comment (arunmk) [inline pkg/testingsdk/k8sclient.go:61]: same: this is also a `waitFor....`

I believe the other ones with the PollImmediate are also like that

File: pkg/testingsdk/k8sclient.go [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+	err := wait.PollImmediate(defaultRetryInterval, defaultRetryTimeout, func() (bool, error) {
+		pods, err := getPodsByPrefix(ctx, k8sClient, nameSpace, deployName)
```
Comment (arunmk) [inline pkg/testingsdk/k8sclient.go:63]: Here we need the pods of a deployment, correct? Those can be done in a different way: https://stackoverflow.com/questions/63142663/using-client-go-api-to-list-pods-managed-by-a-deployment-controller-not-working

**Review by arunmk — APPROVED** [review]
After the changes mentioned it should be good to go.

**Review by ymo24 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)* status: 1 of 4 files reviewed, 13 unresolved discussions (waiting on @arunmk, @lzichong, and @rliang88)

___
*[`pkg/testingsdk/client.go` line 53 at r4](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NJ7cL3sD6C-B3WEqtu5:-NLX8wvc11mBB3OOALlS:b-896fix) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/cb9c8c110e6067cc333a34cf0570e39b662b437a/pkg/testingsdk/client.go#L53)):*
<details><summary><i>Previously, lzichong wrote…</i></summary><blockquote>

Can you replace the codes that doesn't compile due to missing functions/files as TODO for now? This way it doesn't interfere with compiling when others are pulling in the code and submitting their PRs
</blockquote></details>

Done.
___
*[`pkg/testingsdk/client.go` line 68 at r6](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NJ7_6EswYUlwIP9nlAc-r6-68:-NLXMOlvDQXAZ9Igxj-w:b7kw0jq) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/d3be005523816078dd758e2249d9c2be648b11ff/pkg/testingsdk/client.go#L68)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

clusterName would be more helpful in these logs while debugging
</blockquote></details>

Thanks Arun, I have added ClusterName to every log
___
*[`pkg/testingsdk/client.go` line 141 at r6](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NJ7_6EswYUlwIP9nlAc-r6-141:-NLXMSwZEKcNZf8yccBh:b-u8fs2u) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/d3be005523816078dd758e2249d9c2be648b11ff/pkg/testingsdk/client.go#L141)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

error string should be fixed
</blockquote></details>

Thanks Arun, already done
___
*[`pkg/testingsdk/k8sclient.go` line 36 at r6](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIsgS3FNXKOGlMd8a4O-r6-36:-NLXMViVAh2Q1_Avq4X7:b-u8fs2u) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/d3be005523816078dd758e2249d9c2be648b11ff/pkg/testingsdk/k8sclient.go#L36)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

this is actually a `waitForPvcReady`, correct?
</blockquote></details>

Thanks Arun, already done
___
*[`pkg/testingsdk/k8sclient.go` line 61 at r6](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIsgS3FNXKOGlMd8a4O-r6-61:-NLXMY0VBwpBK7mt_HNE:b-896fix) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/d3be005523816078dd758e2249d9c2be648b11ff/pkg/testingsdk/k8sclient.go#L61)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

same: this is also a `waitFor....`

I believe the other ones with the PollImmediate are also like that
</blockquote></details>

Done.
___
*[`pkg/testingsdk/k8sclient.go` line 63 at r6](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183#-NIsgS3FNXKOGlMd8a4O-r6-63:-NLXMYPA68IUKK4GOo5o:b-m1qis) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/d3be005523816078dd758e2249d9c2be648b11ff/pkg/testingsdk/k8sclient.go#L63)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

Here we need the pods of a deployment, correct? Those can be done in a different way: https://stackoverflow.com/questions/63142663/using-client-go-api-to-list-pods-managed-by-a-deployment-controller-not-working
</blockquote></details>

Thanks Arun. I have already fixed it by using labelSelector

**Review by lzichong — APPROVED** [review]
> Reviewed 1 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/183)* status: 2 of 4 files reviewed, 13 unresolved discussions (waiting on @arunmk)

**Pull request merged**